components/uniset/src/uniset.rs

```diff
@@ -0,0 +1,316 @@
+
+impl TryFrom<Vec<u32>> for UnicodeSet {
+    type Error = USetError;
+
+    fn try_from(set: Vec<u32>) -> Result<Self, Self::Error> {
+        if is_valid(&set) {
+            Ok(UnicodeSet { inv_list: set })
+        } else {
+            Err(USetError::InvalidSet(set))
+        }
+    }
+}
+
+impl UnicodeSet {
+    /// Returns UnicodeSet spanning entire Unicode range
+    ///
+    /// The range spans from `0x0 -> 0x10FFFF` inclusive
+    pub fn all() -> UnicodeSet {
+        UnicodeSet {
+            inv_list: vec![0, (MAX as u32) + 1],
+        }
+    }
+
+    /// Returns UnicodeSet spanning BMP range
+    ///
+    /// The range spans from `0x0 -> 0xFFFF` inclusive
+    pub fn bmp() -> UnicodeSet {
+        UnicodeSet {
+            inv_list: vec![0, BMP_MAX + 1],
+        }
+    }
+
+    /// Yields an iterator of start and stop points of ranges in the UnicodeSet
+    ///
+    /// Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::UnicodeSet;
+    /// use std::convert::TryFrom;
+    /// let example_list = vec![0, 10, 15, 20];
+    /// let example = UnicodeSet::try_from(example_list).unwrap();
+    /// let mut example_ranges = example.ranges();
+    /// assert_eq!(Some(&0), example_ranges.next());
+    /// assert_eq!(Some(&10), example_ranges.next());
+    /// assert_eq!(Some(&15), example_ranges.next());
+    /// assert_eq!(Some(&20), example_ranges.next());     
+    /// assert_eq!(None, example_ranges.next());
+    /// ```
+    pub fn ranges(&self) -> Iter<u32> {
+        self.inv_list.iter()
+    }
+
+    /// Yields an iterator going through the character set in the UnicodeSet
+    ///
+    /// Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::UnicodeSet;
+    /// use std::convert::TryFrom;
+    /// let example_list = vec![65, 68, 69, 70];
+    /// let example = UnicodeSet::try_from(example_list).unwrap();
+    /// let mut example_iter = example.iter();
+    /// assert_eq!(Some('A'), example_iter.next());
+    /// assert_eq!(Some('B'), example_iter.next());
+    /// assert_eq!(Some('C'), example_iter.next());
+    /// assert_eq!(Some('E'), example_iter.next());
+    /// assert_eq!(None, example_iter.next());
+    /// ```
+    pub fn iter(&self) -> impl Iterator<Item = char> + '_ {
+        self.inv_list
+            .chunks(2)
+            .flat_map(|pair| (pair[0]..pair[1]))
+            .filter_map(from_u32)
+    }
+
+    /// Returns the number of elements of the UnicodeSet
+    pub fn size(&self) -> usize {
+        if self.is_empty() {
+            return 0;
+        }
+        let s: u32 = self
+            .inv_list
+            .chunks(2)
+            .map(|end_points| end_points[1] - end_points[0])
+            .sum();
+        s as usize
+    }
+
+    /// Returns whether or not the UnicodeSet is empty
+    pub fn is_empty(&self) -> bool {
+        self.inv_list.is_empty()
+    }
+
+    /// Wrapper for contains
+    ///
+    /// Returns an Option as to whether or not it is possible for the query to be contained
```

Comment:
Clarify that the number inside the `Option` is the array index of the _start of the range_ that contains the query.

Reply:
agreed and changed